### ISSUE TITLE
Default function argument is not assigned to TypeVar

### ISSUE THREAD

**Issue by dsimon-dev** [microsoft/pyright#2561]
Consider the following typed utility function:

```python
T = TypeVar("T")
TDefault = TypeVar("TDefault")

def at(ls: list[T], i: int, default: TDefault = None) -> Union[T, TDefault]:
    try:
        return ls[i]
    except IndexError:
        return default
```

Pylance reports an error at the `return default` line:

```
Expression of type "TDefault@at | None" cannot be assigned to return type "T@at | TDefault@at"
  Type "object* | None" cannot be assigned to type "T@at | TDefault@at"
    Type "None" cannot be assigned to type "T@at | TDefault@at"
      Type "None" cannot be assigned to type "T@at"
      Type "None" cannot be assigned to type "TDefault@at" Pylance(reportGeneralTypeIssues)
```

Why is the type of `default` `TDefault@at | None` instead of just `TDefault@at`? In my opinion the type of a default value should be forced to be assignable to the argument type, and not automatically added as a union type.

Is there a workaround to correctly type this function?

**Comment by erictraut** [comment]
The problem here is that a parameter with a default value of `None` is considered shorthand for an optional type. This is an old convention that existing code often relies on, but it's unfortunate because it leaves room for ambiguity about which type you really intend.

```python
def func(a: str = None) ...
```

Is equivalent to:

```python
def func(a: Optional[str] = None) ...
```

Pyright has a mode where you can turn off this "feature" and require an explicit `Optional`: `strictParameterNoneValue`.

You can enable this strict flag for your entire project or on a file-by-file basis using the following comment at the top of the file:

```python
# pyright: strictParameterNoneValue=true
```

PEP 484 used to require that type checkers treat `= None` as an implied `Optional` type, but the PEP has since been modified to say:
> This is no longer the recommended behavior. Type checkers should move towards requiring the optional type to be made explicit.

So perhaps it's time that we consider changing the default setting for `strictParameterNoneValue` and turn it on by default.

**Comment by dsimon-dev** [comment]
Thanks for the explanation! I have created a `pyrightconfig.json` file and added `"strictParameterNoneValue": true` which fixed this issues.
